### PR TITLE
busybox: Add a fix for CVE-2021-28831

### PR DESCRIPTION
**Pull request by TethysSvensson**
###### Motivation for this change

It's a security fix, see #120362 and #120363

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 121578` at 46c2d4c5 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>nix-exec</li>
    <li>nixui</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/a392acbdbbdd37c12a292aed7e2f9399">dydisnix</a></li>
    <li><a href="https://gist.github.com/b413dce3163fc567415d290247d85932">libnixxml</a></li>
  </ul>
</details>
<details>
  <summary>63 packages built successfully:</summary>
  <ul>
    <li>bundix</li>
    <li>busybox</li>
    <li>busybox-sandbox-shell</li>
    <li>cabal2nix</li>
    <li>cachix</li>
    <li>crate2nix</li>
    <li>dep2nix</li>
    <li>disnix</li>
    <li>disnixos</li>
    <li>epson-escpr2</li>
    <li>fcron</li>
    <li>fusionInventory</li>
    <li>go2nix</li>
    <li>haskellPackages.cachix</li>
    <li>haskellPackages.hercules-ci-agent</li>
    <li>haskellPackages.hercules-ci-cnix-expr</li>
    <li>haskellPackages.hercules-ci-cnix-store</li>
    <li>haskellPackages.nix-paths</li>
    <li>haskellPackages.update-nix-fetchgit</li>
    <li>hercules-ci-agent</li>
    <li>hydra-unstable</li>
    <li>lazpaint</li>
    <li>lightdm</li>
    <li>lightdm-enso-os-greeter</li>
    <li>lightdm-mini-greeter</li>
    <li>lightdm-tiny-greeter</li>
    <li>lightdm_gtk_greeter</li>
    <li>lightdm_qt</li>
    <li>molly-guard</li>
    <li>nix (nixStable)</li>
    <li>nix-bundle</li>
    <li>nix-direnv</li>
    <li>nix-doc</li>
    <li>nix-du</li>
    <li>nix-index</li>
    <li>nix-pin</li>
    <li>nix-plugins</li>
    <li>nix-prefetch-bzr</li>
    <li>nix-prefetch-cvs</li>
    <li>nix-prefetch-docker</li>
    <li>nix-prefetch-git</li>
    <li>nix-prefetch-hg</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-prefetch-svn</li>
    <li>nix-serve</li>
    <li>nix-update</li>
    <li>nix-update-source</li>
    <li>nixFlakes (nixUnstable)</li>
    <li>nixStatic</li>
    <li>nixos-generators</li>
    <li>nixos-shell</li>
    <li>nixpkgs-review</li>
    <li>pantheon.elementary-greeter</li>
    <li>pipework</li>
    <li>python38Packages.nix-kernel</li>
    <li>python38Packages.nixpkgs</li>
    <li>python38Packages.pythonix</li>
    <li>python39Packages.nix-kernel</li>
    <li>python39Packages.nixpkgs</li>
    <li>python39Packages.pythonix</li>
    <li>update-nix-fetchgit</li>
    <li>vgo2nix</li>
    <li>vulnix</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, please check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/os-specific/linux/busybox/default.nix:141:5:
```
    |
141 |     license = licenses.gpl2;
    |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`configurePhase` should probably contain `runHook preConfigure`.

Near pkgs/os-specific/linux/busybox/default.nix:75:3:
```
   |
75 |   configurePhase = ''
   |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement.

**Comment by SuperSandro2000** [comment]
Where did you get the patch from? Can we fetchpatch it?

**Comment by TethysSvensson** [comment]
I got it here: https://git.busybox.net/busybox/commit/?id=f25d254dfd4243698c31a4f3153d4ac72aa9e9bd

I pushed a version that uses fetchpatch.

File: pkgs/os-specific/linux/busybox/default.nix [hunk 1]
```diff
@@ -64,6 +67,10 @@ stdenv.mkDerivation rec {
 
   patches = [
     ./busybox-in-store.patch
+    (fetchpatch {
+      url = "https://git.busybox.net/busybox/patch/?id=f25d254dfd4243698c31a4f3153d4ac72aa9e9bd";
```
Comment (mweinelt) [inline pkgs/os-specific/linux/busybox/default.nix:71]:
```suggestion
      name = "CVE-2021-28831.patch";
      url = "https://git.busybox.net/busybox/patch/?id=f25d254dfd4243698c31a4f3153d4ac72aa9e9bd";
```

Giving the patch a name makes tracking fixed CVEs easier.

**Review by mweinelt — CHANGES_REQUESTED** [review]

**Review by mweinelt — APPROVED** [review]

**Comment by mweinelt** [comment]
@TethysSvensson Can you backport this?

**Comment by TethysSvensson** [comment]
@mweinelt Yes, though I do not feel too comfortable about it. 20.09 is still running 1.31.1, and the patch does not apply cleanly. I think I could backport the patch, but I would not be sure if I was missing context. Perhaps we could bump 20.09 to use 1.32.1 as well?

**Pull request merged**